packages/jsii-diff/lib/type-analysis.ts

```diff
@@ -0,0 +1,171 @@
+      ...flatMap(analyses, x => x.success ? [] : x.reasons)
+    );
+  }
+
+  // For named types, we'll always do a nominal typing relationship.
+  // That is, if in the updated typesystem someone were to use the type name
+  // from the old assembly, do they have a typing relationship that's accepted
+  // by a nominal type system. (That check also rules out enums)
+  const nominalCheck = isNominalSuperType(a, b, updatedSystem);
+  if (nominalCheck.success === false) { return nominalCheck; }
+
+  // At this point, the types are legal in the updated assembly's type system.
+  // However, for structs we also structurally check the fields between the OLD
+  // and the NEW type system.
+  // We could do more complex analysis on typing of methods, but it doesn't seem
+  // worth it.
+  const A = a.type!; // Note: lookup in old type system!
+  const B = b.type!;
+  if (A.isInterfaceType() && A.isDataType() && B.isInterfaceType() && B.datatype) {
+    return isStructuralSuperType(A, B, updatedSystem);
+  }
+
+  // All seems good
+  return { success: true };
+}
+
+/**
+ * Find types A and B in the updated type system, and check whether they have a supertype relationship in the type system
+ */
+function isNominalSuperType(a: reflect.TypeReference, b: reflect.TypeReference, updatedSystem: reflect.TypeSystem): Analysis {
+  if (a.fqn === undefined) {
+    throw new Error(`I was expecting a named type, got '${a}'`);
+  }
+
+  // Named type vs a non-named type
+  if (b.fqn === undefined) { return failure(`${b} is not assignable to ${a}`); }
+
+  // Short-circuit of the types are the same name, saves us some lookup
+  if (a.fqn === b.fqn) { return { success: true }; }
+
+  // We now need to do subtype analysis on the
+  // Find A in B's typesystem, and see if B is a subtype of A'
+  const B = updatedSystem.tryFindFqn(b.fqn!);
+  const A = updatedSystem.tryFindFqn(a.fqn);
+
+  if (!B) { return failure(`could not find type ${b}`); }
+  if (!A) { return failure(`could not find type ${a}`); }
+
+  // If they're enums, they should have been exactly the same (tested above)
+  // enums are never subtypes of any other type.
+  if (A.isEnumType()) { return failure(`${a} is an enum different from ${b}`); }
+  if (B.isEnumType()) { return failure(`${b} is an enum different from ${a}`); }
+
+  if (B.extends(A)) { return { success: true }; }
+  return failure(`${b} does not extend ${a}`);
+}
+
+/**
+ * Is struct A a structural supertype of struct B?
+ *
+ * Trying to answer the question, is this assignment legal for all values
+ * of `expr in B`.
+ *
+ * ```ts
+ * const var: A = expr as B;
+ * ```
+ *
+ * A is a structural supertype of B if all required members of A are also
+ * required in B, and of a compatible type.
+ *
+ * Optional members of A must either not exist in B, or be of a compatible
+ * type.
+ */
+function isStructuralSuperType(a: reflect.InterfaceType, b: reflect.InterfaceType, updatedSystem: reflect.TypeSystem): Analysis {
```

Comment:
I am tempted to suggest that you modify reflect to represent structs separately from
Interfaces...